crates/ruff_python_formatter/src/statement/stmt_with.rs

```diff
@@ -130,92 +132,167 @@ impl FormatNodeRule<StmtWith> for FormatStmtWith {
     }
 }
 
-/// Determines whether the `with` items should be parenthesized (over parenthesizing each item),
-/// and if so, which parenthesizing layout to use.
-///
-/// Parenthesize `with` items if
-/// * The last item has a trailing comma (implying that the with items were parenthesized in the source)
-/// * There's more than one item and they're already parenthesized
-/// * There's more than one item, the [`wrap_multiple_context_managers_in_parens`](is_wrap_multiple_context_managers_in_parens) preview style is enabled,
-///     and the target python version is >= 3.9
-/// * There's a single non-parenthesized item. The function returns [`ParenthesizeWith::Optional`]
-///     if the parentheses can be omitted if breaking around parenthesized sub-expressions is sufficient
-///     to make the expression fit. It returns [`ParenthesizeWith::IfExpands`] otherwise.
-/// * The only item is parenthesized and has comments.
-fn should_parenthesize(
-    with: &StmtWith,
-    options: &PyFormatOptions,
-    context: &PyFormatContext,
-) -> FormatResult<ParenthesizeWith> {
-    if has_magic_trailing_comma(with, options, context) {
-        return Ok(ParenthesizeWith::IfExpands);
-    }
+#[derive(Clone, Copy, Debug)]
+enum WithItemsLayout<'a> {
+    /// The with statement's only item has a parenthesized context manager.
+    ///
+    /// ```python
+    /// with (
+    ///     a + b
+    /// ):
+    ///     ...
+    ///
+    /// with (
+    ///     a + b
+    /// ) as b:
+    ///     ...
+    /// ```
+    ///
+    /// In this case, prefer keeping the parentheses around the context expression instead of parenthesizing the entire
+    /// with item.
+    SingleParenthesizedContextManager(&'a WithItem),
 
-    let can_parenthesize = options.target_version() >= PythonVersion::Py39
-        || are_with_items_parenthesized(with, context)?;
+    /// It's a single with item with a target. Use the optional parentheses layout (see [`optional_parentheses`])
+    /// to mimic the `maybe_parenthesize_expression` behavior.
+    ///
+    /// ```python
+    /// with (
+    ///     a + b
+    /// ) as b:
+    ///     ...
+    /// ```
+    ///
+    /// Only used for Python 3.9+
+    SingleWithTarget(&'a WithItem),
```

Comment:
The name here doesn't make sense as of this PR because it could as well be a with item without a target. This gets changed in https://github.com/astral-sh/ruff/pull/10274, and I didn't want to come up with new terminology just for the state between the two PRs.